lighthouse-core/gather/computed/first-interactive.js

```diff
@@ -0,0 +1,185 @@
+    const tInSeconds = t / 1000;
+    const exponentiationComponent = Math.exp(-.045 * tInSeconds);
+    return (4 * exponentiationComponent + 1) * 1000;
+  }
+
+  /**
+   * @param {!Array<{start: number, end: number>}} longTasks
+   * @param {number} i
+   * @param {number} FMP
+   * @param {number} traceEnd
+   * @return {number} The last index of the lonely task envelope,
+   *    -1 if not a lonely task
+   */
+  static getLastLonelyTaskIndex(longTasks, i, FMP, traceEnd) {
+    const longTask = longTasks[i];
+    const previousTask = longTasks[i - 1];
+    if (longTask.start < FMP + LONELY_TASK_FMP_DISTANCE ||
+        traceEnd - longTask.end < LONELY_TASK_NEIGHBOR_DISTANCE ||
+        longTask.start - previousTask.end < LONELY_TASK_NEIGHBOR_DISTANCE ||
+        longTask.end - longTask.start > LONELY_TASK_ENVELOPE_SIZE) {
+      return -1;
+    }
+
+    let lonelyTaskIndex = i;
+    const envelopeEnd = longTask.start + LONELY_TASK_ENVELOPE_SIZE;
+    const windowEnd = envelopeEnd + LONELY_TASK_NEIGHBOR_DISTANCE;
+    for (let j = i + 1; longTasks[j] && longTasks[j].start < windowEnd; j++) {
+      if (longTasks[j].end > envelopeEnd) {
+        lonelyTaskIndex = -1;
+      } else {
+        lonelyTaskIndex = j;
+      }
+    }
+
+    return lonelyTaskIndex;
+  }
+
+  /**
+   * @param {number} FMP
+   * @param {number} traceEnd
+   * @param {!Array<{start: number, end: number>}} longTasks
+   * @return {number}
+   */
+  static findQuietWindow(FMP, traceEnd, longTasks) {
+    if (longTasks.length === 0 ||
+        longTasks[0].start > FMP + MAX_QUIET_WINDOW_SIZE) {
+      return FMP;
+    }
+
+    // FirstInteractive must start at the end of a long task, consider each long task and
+    // examine the window that follows it.
+    for (let i = 0; i < longTasks.length; i++) {
+      const event = longTasks[i];
+      const windowStart = event.end;
+      const windowSize = FirstInteractive.getRequiredWindowSizeInMs(windowStart - FMP);
+      const windowEnd = windowStart + windowSize;
+
+      // Check that we have a long enough trace
+      if (windowEnd > traceEnd) {
+        throw new Error(TRACE_BUSY_MSG);
+      }
+
+      let isQuiet = true;
+      // Loop over all the long tasks within the window
+      // All tasks that we find in the window must be lonely or the window isn't quiet.
+      for (let j = i + 1; j < longTasks.length && longTasks[j].start < windowEnd; j++) {
+        const lastLonelyTaskIndex = FirstInteractive.getLastLonelyTaskIndex(longTasks, j, FMP,
+            traceEnd);
+        if (lastLonelyTaskIndex === -1) {
+          // We found a task that isn't lonely, this window isn't quiet
+          isQuiet = false;
+          break;
+        } else {
+          // Skip over the rest of the lonely task envelope
+          j = lastLonelyTaskIndex;
+        }
+      }
+
+      if (isQuiet) {
+        return windowStart;
+      }
+    }
+
+    throw new Error(TRACE_BUSY_MSG);
+  }
+
+  /**
+   * @param {!Object} trace
```

Comment:
![image](https://cloud.githubusercontent.com/assets/2301202/25396106/2cdf7d30-2998-11e7-89c0-23f734080b55.png)

😆  sorry I had to, seriously though it'd be nice to have a conversation about what we actually care about typing and what we don't, these are all computed artifacts that do not make sense writing one-off types for in every place that uses them but also would give no value to the developer as documentation if they were typed somewhere else and the type was just `!Trace` or something equally useless that just matched the name of the parameter.